native/dsterf.go

```diff
@@ -55,241 +55,222 @@ func (impl Implementation) Dsterf(n int, d, e []float64) (ok bool) {
-			d[l], d[l+1] = impl.Dlae2(d[l], rte, d[l+1])
-			e[l] = 0
-			l += 2
-			if l <= lend {
-				goto Fifty
+		var m int
+		for m = l1; m < n-1; m++ {
+			if math.Abs(e[m]) <= math.Sqrt(math.Abs(d[m]))*math.Sqrt(math.Abs(d[m+1]))*eps {
+				e[m] = 0
+				break
 			}
-			goto OneFifty
-		}
-		if jtot == nmaxit {
-			goto OneFifty
 		}
-		jtot++
 
-		// Form shift.
-		rte = math.Sqrt(e[l])
-		sigma = (d[l+1] - p) / (2 * rte)
-		r = impl.Dlapy2(sigma, 1)
-		sigma = p - (rte / (sigma + math.Copysign(r, sigma)))
-
-		c = 1
-		s = 0
-		gamma = d[m] - sigma
-		p = gamma * gamma
-
-		// Inner loop.
-		for i := m - 1; i >= l; i-- {
-			bb := e[i]
-			r := p + bb
-			if i != m-1 {
-				e[i+1] = s * r
-			}
-			oldc := c
-			c = p / r
-			s = bb / r
-			oldgam := gamma
-			alpha := d[i]
-			gamma = c*(alpha-sigma) - s*oldgam
-			d[i+1] = oldgam + (alpha - gamma)
-			if c != 0 {
-				p = (gamma * gamma) / c
-			} else {
-				p = oldc * bb
-			}
+		l := l1
+		lsv := l
+		lend := m
+		lendsv := lend
+		l1 = m + 1
+		if lend == 0 {
+			continue
 		}
-		e[l] = s * p
-		d[l] = sigma + gamma
-		goto Fifty
 
-	Ninety:
-		// Eigenvalue found.
-		d[l] = p
-		l++
-		if l <= lend {
-			goto Fifty
-		}
-		goto OneFifty
-	} else {
-	OneHundred:
-		// QR Iteration.
-		// Look for small super-diagonal element.
-		for m = l; m >= lend+1; m-- {
-			if math.Abs(e[m-1]) <= eps2*math.Abs(d[m]*d[m-1]) {
-				goto OneTwenty
-			}
-		}
-		m = lend
-	OneTwenty:
-		if m > lend {
-			e[m-1] = 0
+		// Scale submatrix in rows and columns l to lend.
+		anorm := impl.Dlanst(lapack.MaxAbs, lend-l+1, d[l:], e[l:])
+		iscale := none
+		if anorm == 0 {
+			continue
 		}
-		p := d[l]
-		if m == l {
-			goto OneFourty
+		if anorm > ssfmax {
+			iscale = down
+			impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l+1, 1, d[l:], n)
+			impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l, 1, e[l:], n)
+		} else if anorm < ssfmin {
+			iscale = up
+			impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l+1, 1, d[l:], n)
+			impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l, 1, e[l:], n)
 		}
 
-		// If remaining matrix is 2 by 2, use Dlae2 to compute its eigenvalues.
-		if m == l-1 {
-			rte = math.Sqrt(e[l-1])
-			d[l], d[l-1] = impl.Dlae2(d[l], rte, d[l-1])
-			e[l-1] = 0
-			l -= 2
-			if l >= lend {
-				goto OneHundred
-			}
-			goto OneFifty
+		el := e[l:lend]
+		for i, v := range el {
+			el[i] *= v
 		}
-		if jtot == nmaxit {
-			goto OneFifty
+
+		// Choose between QL and QR iteration.
+		if math.Abs(d[lend]) < math.Abs(d[l]) {
+			lend = lsv
+			l = lendsv
 		}
-		jtot++
+		if lend >= l {
+			// QL Iteration.
+			// Look for small sub-diagonal element.
+			for {
+				if l != lend {
+					for m = l; m < lend; m++ {
+						if math.Abs(e[m]) <= eps2*(math.Abs(d[m]*d[m+1])) {
+							break
+						}
+					}
+				} else {
+					m = lend
+				}
+				if m < lend {
+					e[m] = 0
+				}
+				p := d[l]
+				if m == l {
+					// Eigenvalue found.
+					l++
+					if l <= lend {
```

Comment:
In all of the other files, the idiom was changed to have the break case first. That should probably be done here as well.